dao-voting/tests/harness.rs

```diff
@@ -0,0 +1,898 @@
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user, 10, 10, proposal_transaction.clone()).await;
+            vote(&user, true, 0, 0).await;
+        }
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_on_expired_proposal() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+            constructor(&deployer, gov_token_id).await;
+
+            mint(
+                &deployer.gov_token.as_ref().unwrap(),
+                100,
+                user.wallet.address(),
+            )
+            .await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user, 1, 1, proposal_transaction.clone()).await;
+
+            let tx_params = TxParameters::new(None, Some(1_000_000), None, None);
+            let call_params = CallParameters::new(
+                Some(asset_amount),
+                Some(AssetId::from(*gov_token_id)),
+                Some(100_000),
+            );
+            deposit(&user, tx_params, call_params).await;
+            vote(&user, true, 0, asset_amount / 4).await;
+        }
+
+        #[tokio::test]
+        #[should_panic]
+        async fn panics_on_vote_amount_greater_than_balance() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+            constructor(&deployer, gov_token_id).await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user, 10, 10, proposal_transaction.clone()).await;
+            vote(&user, true, 10, asset_amount).await;
+        }
+    }
+}
+
+mod execute {
+    use super::*;
+
+    mod success {
+        use super::*;
+
+        // TODO: uncomment this test once more support is added for the CALL opcode
+        #[tokio::test]
+        #[ignore]
+        async fn user_proposal_can_execute() {
+            let (_gov_token, gov_token_id, deployer, user, asset_amount) = setup().await;
+            constructor(&deployer, gov_token_id).await;
+
+            mint(
+                &deployer.gov_token.as_ref().unwrap(),
+                100,
+                user.wallet.address(),
+            )
+            .await;
+
+            let tx_params = TxParameters::new(None, Some(1_000_000), None, None);
+            let call_params = CallParameters::new(
+                Some(asset_amount),
+                Some(AssetId::from(*gov_token_id)),
+                Some(100_000),
+            );
+            deposit(&user, tx_params, call_params).await;
+
+            let proposal_transaction = proposal_transaction(gov_token_id);
+            create_proposal(&user, 10, 1, proposal_transaction.clone()).await;
+            vote(&user, true, 0, asset_amount / 2).await;
+
+            execute(&user, 0).await;
+
+            // TODO actually test execution of an arbitrary transaction
+        }
+    }
+
+    mod revert {
```

Comment:
Another test to consider is whether we have a revert for a failed call (once that works)